info_extraction/src/edu/gatech/gtri/info_extraction/utils.py

```diff
@@ -0,0 +1,125 @@
+from collections import defaultdict
+from functools import reduce
+from bs4 import BeautifulSoup
+from markdown import markdown
+from pathlib import Path
+import re
+
+
+# gist:https://gist.github.com/lorey/eb15a7f3338f959a78cc3661fbc255fe
+def markdown_to_text(markdown_string):
+    """ Converts a markdown string to plaintext """
+
+    # md -> html -> text since BeautifulSoup can extract text cleanly
+    html = markdown(markdown_string)
+
+    # remove code snippets
+    html = re.sub(r'<pre>(.*?)</pre>', ' ', html)
+    html = re.sub(r'<code>(.*?)</code >', ' ', html)
+
+    # extract text
+    soup = BeautifulSoup(html, "html.parser")
+    text = ''.join(soup.findAll(text=True))
+
+    return text
+
+
+"""
+Take an array of julia code lines and extract comments
+"""
+
+
+def julia_comment_extract(julia_code_lines):
+    keeps = []
+    for line in julia_code_lines:
+        processed_line = line.lstrip(" ")
+        if processed_line.startswith("#"):
+            keeps.append(line)
+
+    return keeps
+
+
+"""
+Take an array of julia code lines and extract parameters from julia function declarations
+"""
+
+
+def julia_param_extract(julia_code_lines):
+    keeps = []
+    func_match = r'function\s*(.*?)\((.*?)\)'
+    for line in julia_code_lines:
+
+        match = re.search(func_match, line)
+        if match:
+            grps = match.groups()
+            func_name = match.group(1)
+            params = match.group(2)
+            func_pair = [func_name, params.split(",")]
+            keeps.append(func_pair)
+
+    return keeps
+
+
+"""
+Associate tokens from comments with parameters
+"""
+
+
+def intersect_comments_params(comment_lines, param_pairs):
+    # :comment_lines are textual comments extract from julia code
+    # :param_lines are arrays of parameters extracted from each and every function signature along with function name
```

Comment:
This says `:param_lines` but the actual parameter is `param_pairs`. Docstring needs to be updated.

From the code it looks like `param_pairs` is a list `[(k, [p1,p2,...])]` is that right? What is `k`, and the `pi`?

Reply:
I noticed the colon-var syntax in Julia and was just using it here;
k is the function name and pi is a list of extracted parameters